iOS/app/AppDelegate.swift

```diff
@@ -0,0 +1,192 @@
+import HTMLString
+import os
+import UIKit
+
+// When an app update ships with a new database, either because the dictionary
+// was updated or migrations were added, we need to copy it into place. But
+// we don’t want to do that on every launch because it would be slow.
+//
+// So the code will check whether the version number saved into a file matches.
+//
+// This could be automated by comparing query results between the database
+// file shipped with the app and the writable copy in the “Application Support”
+// directory.
+let CURRENT_DB_VERSION = "1"
+
+let log = OSLog(subsystem: "morphodict", category: "default")
+
+var appDelegate: AppDelegate?
+var serverHasBeenPreviouslyStarted = false
+
+func serveCallback() {
+    if !serverHasBeenPreviouslyStarted {
+        DispatchQueue.main.async {
+            appDelegate?.viewController?.goHome()
+        }
+    }
+    serverHasBeenPreviouslyStarted = true
+}
+
+@main
+class AppDelegate: UIResponder, UIApplicationDelegate {
+    var window: UIWindow?
+    var viewController: ViewController?
+    private let fm = FileManager.default
+    private let applicationSupportDirectory: URL
+
+    override init() {
+        applicationSupportDirectory = try! fm.url(
+            for: .applicationSupportDirectory, in: .userDomainMask,
+            appropriateFor: nil, create: true)
+
+        super.init()
+    }
+
+    func application(_: UIApplication, didFinishLaunchingWithOptions _:
+        [UIApplication.LaunchOptionsKey: Any]?) -> Bool
+    {
+        window = window ?? UIWindow()
+        appDelegate = self
+        viewController = ViewController()
+        window!.rootViewController = viewController
+        window!.makeKeyAndVisible()
+
+        setenv("MORPHODICT_ENV_FILE_PATH",
+               applicationSupportDirectory.appendingPathComponent(".env").path,
+               1)
+        morphodict_register_callback("serve", serveCallback)
+
+        let pyq = DispatchQueue(
+            label: "python server", qos: .userInitiated, attributes: [],
+            autoreleaseFrequency: .inherit, target: .none)
+        pyq.async {
+            morphodict_register_callback("hello") { print("Hello from swift") }
+
+            do {
+                try self.setupDatabase()
+            } catch {
```

Comment:
Does this need to catch the error? As in, `catch error {...`

Reply:
Good question! Nope, in swift, while you can adjust the behaviour, it is implied that you are catching something into a var named `error`

https://docs.swift.org/swift-book/LanguageGuide/ErrorHandling.html#ID541
> You write a pattern after `catch` to indicate what errors that clause can
> handle. If a `catch`clause doesn’t have a pattern, the clause matches any
> error and binds the error to a local constant named `error`.